python_modules/dagster/dagster/_core/definitions/metadata/source_code.py

```diff
@@ -142,25 +142,59 @@ def _with_code_source_single_definition(
     )
 
 
-def convert_local_path_to_source_control_path(
-    base_source_control_url: str,
-    repository_root_absolute_path: str,
+FilePathMappingFn = Callable[[Path], str]
+
+
+class AnchorBasedFilePathMappingFn(NamedTuple):
+    """Specifies the mapping between local file paths and their corresponding paths in a source control repository,
+    using a specific anchor file as a reference point. All other paths are calculated relative to this anchor file.
+
+    Args:
+        local_file_anchor (Path): The path to a local file that is present in the repository.
+        file_anchor_path_in_repository (str): The path to the anchor file in the repository.
+
+    Example:
+        .. code-block:: python
+
+            mapping_fn = AnchorBasedFilePathMappingFn(
+                local_file_anchor=Path(__file__),
+                file_anchor_path_in_repository="python_modules/my_module/my-module/__init__.py",
+            )
+    """
+
+    local_file_anchor: Path
+    file_anchor_path_in_repository: str
+
+    def __call__(self, local_path: Path) -> str:
+        path_from_anchor_to_target = os.path.relpath(
+            local_path,
+            self.local_file_anchor,
+        )
+        return os.path.normpath(
+            os.path.join(
+                self.file_anchor_path_in_repository,
+                path_from_anchor_to_target,
+            )
+        )
```

Comment:
Like @sryza I'm skeptical of overloading `__call__` here. Can we just have a vanilla method?

Reply:
We can, it just means the user can't provide a lambda here which is a bit of a bummer. Maybe we expect that use-case to be uncommon enough that it's fine?

Reply:
For now, updated to an explicit ABC with non-`__call__` method